lib/qos/qos.c

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <sys/slist.h>
+#include <qos.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(qos, CONFIG_QOS_LOG_LEVEL);
+
+/* Structure used to keep track of pending messages. Used in combination with a linked list. */
+struct qos_metadata {
+	/* Mandatory variable used to construct a linked list. */
+	sys_snode_t header;
+
+	/* Message associated with each node in the linked list. */
+	struct qos_data message;
+};
+
+/* Structure containing internal variables in the library.  */
+static struct ctx {
+	/* Library event handler. Used in callbacks to the caller. */
+	qos_evt_handler_t app_evt_handler;
+
+	/* Internal array of pending messages. Used in combination with linked list. */
+	struct qos_metadata list_internal[CONFIG_QOS_PENDING_MESSAGES_MAX];
+
+	/* Linked list used to keep track of pending messages. */
+	sys_slist_t pending_list;
+
+	/* Variable used to prevent multiple library initializations. */
+	bool initialized;
+
+	/* Delayed work used as the backoff timer. */
+	struct k_work_delayable timeout_handler_work;
+} ctx;
+
+/* Mutex used to protect internal library variables. */
+static K_MUTEX_DEFINE(ctx_lock);
```

Comment:
I think the mutex could be part of the context struct to keep things consistent

Reply:
I tried doing this, but the problem is that the mutex needs to be initialized at run-time and I cannot initialize the mutex in the `qos_init`. Since the lock is being used there as well. I could use SYS_INIT to initialize the mutex but I'm not convinced its a good tradeoff to get a little more consistency.